### PR TITLE
Scalable neural network.

### PR DESCRIPTION
**Pull request by sopeal**

### CODE REVIEW

File: NeuralNetwork.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+import random
+from scipy.special import expit # fast logistic function
```
Comment (agaas) [inline NeuralNetwork.py:2]: Please add info to readme about python version that should be used and required libraries.

File: runExample.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+iterations = 10000
+history_errors = []
+for i in range(iterations):
+    training_inputs, training_outputs = random.choice(training_sets)
+    nn.train(training_inputs, training_outputs)
+    history_errors.append(nn.calculate_total_error(training_sets))
+
+
+plt.plot(history_errors)
+plt.show()
```
Comment (agaas) [inline runExample.py]: Please add title and x, y axes descriptions.

File: runExample.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+print(nn.feed_forward(training_sets[0][0]))
+print(nn.feed_forward(training_sets[1][0]))
+print(nn.feed_forward(training_sets[2][0]))
+print(nn.feed_forward(training_sets[3][0]))
```
Comment (agaas) [inline runExample.py]: Missing new line.

File: runExample.py (hunk 2 above)
Comment (agaas) [inline runExample.py]: Function show is blocking executing the rest of the program, in this case you can move show() to the end of the file.

File: NeuralNetwork.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        output_de_dz = [0] * len(self.layers[-1].neurons)
+        for o in range(len(self.layers[-1].neurons)):
+            output_de_dz[o] = self.layers[-1].neurons[o].de_dz(training_outputs[o])
+        self.layers_de_dz[-1] = output_de_dz
+
+    def calculate_hidden_de_dz(self):
+        for i in range(2, len(self.layers) + 1):
+            hidden_de_dz = [0] * len(self.layers[-i].neurons)
+            for h in range(len(self.layers[-i].neurons)):
+
```
Comment (agaas) [inline NeuralNetwork.py]: This empty line is confusing, please remove it.

File: NeuralNetwork.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    def __init__(self, num_neurons):
+        self.bias = random.random()
+        self.initialize_neurons(num_neurons)
+
+    def initialize_neurons(self, num_neurons):
+        self.neurons = []
+        for i in range(num_neurons):
+            self.neurons.append(Neuron(self.bias))
+
+    def feed_forward(self, inputs):
```
Comment (agaas) [inline NeuralNetwork.py]: Body of this function can be substituted with on line (using list comprehensions):
return [neuron.calculate_output(inputs) for neuron in self.neurons]

File: NeuralNetwork.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                total_error += self.layers[-1].neurons[o].error(training_outputs[o])
+        return total_error
+
+
+class NeuronLayer:
+    def __init__(self, num_neurons):
+        self.bias = random.random()
+        self.initialize_neurons(num_neurons)
+
+    def initialize_neurons(self, num_neurons):
```
Comment (agaas) [inline NeuralNetwork.py]: List comprehensions can be used in this function.

File: NeuralNetwork.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+class NeuralNetwork:
+
+    def __init__(self, size):
+        self.LEARNING_RATE = 0.5
+        self.layers = []
+        self.init_layers(size)
+        self.init_weights(size)
+
+    def init_layers(self, size):
+        for i in range(1, len(size)):
```
Comment (agaas) [inline NeuralNetwork.py]: Simpler would be:
self.layers = [NeuronLayer(layer_size) for layer_size in size[1:]]

File: NeuralNetwork.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            self.layers.append(NeuronLayer(size[i]))
+
+    def init_weights(self, size):
+        for j in range(1, len(size)):
+            for h in range(size[j]):
+                for i in range(size[j-1]):
```
Comment (agaas) [inline NeuralNetwork.py]: You can substitute last for function with list comprehension. 
There is no use in naming unused value in for function, you could write:
for _ in range(size[j-1]):

File: NeuralNetwork.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                    self.layers[j-1].neurons[h].weights.append(random.random())
+
+    def feed_forward(self, inputs):
+        output = inputs
+        for i in range(len(self.layers)):
```
Comment (agaas) [inline NeuralNetwork.py]: Instead of this function body you could write one line list comprehension.

File: NeuralNetwork.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+class Neuron:
+    def __init__(self, bias):
+        self.bias = bias
+        self.weights = []
+
+    def calculate_output(self, inputs):
+        self.inputs = inputs
+        total = 0
+        for i in range(len(self.inputs)):
```
Comment (agaas) [inline NeuralNetwork.py]: This code is OK but could be write in few ways, for example:
1) you could iterate over two lists
for input, weight in zip(self.inputs, self.weights):
    total += input * weight
2) you could use list comprehensions to create list of multiplied values, and sum it with build in function:
total = [input * weight for input, weight in zip(self.inputs, self.weights)]
total = sum(total)
3) you could use function from e.g. numpy package to multiply two lists (to be checked)

File: NeuralNetwork.py (hunk 9 above)
Comment (sopeal) [reply NeuralNetwork.py]: I don't know how to do it.

**Pull request merged**